app/views/govuk_component/task_list.raw.html.erb

```diff
@@ -17,70 +19,132 @@
+                            link_href = link[:href]
+                            link_index += 1
+                          end
+
+                          if link[:active]
+                            link_href = "#content"
+                            link_text = '<span class="visuallyhidden">You are currently viewing:</span> '.html_safe + link_text
+                          end
+                        %>
+                        <li class="pub-c-task-list__link">
+                          <% if link[:href] %>
+                            <a href="<%= link_href %>"
+                              class="pub-c-task-list__link-item js-link <% if link[:active] %>pub-c-task-list__link-item--active<% end %>"
+                              data-position="<%= "#{group_index + 1}.#{step_index + 1}.#{link_index}" %>">
+                          <% end %>
+
+                          <%= link_text %>
 
-                <% if step[:panel_links] %>
-                  <ol class="pub-c-task-list__panel-links" data-length="<%= step[:panel_links].length %>">
-                    <% step[:panel_links].each_with_index do |panel_link, panel_index| %>
-                      <%
-                        panel_link_href = panel_link[:href]
-                        panel_link_text = ""
-
-                        if panel_link[:active]
-                          panel_link_href = "#content"
-                          panel_link_text = '<span class="visuallyhidden">You are currently viewing:</span>'.html_safe
-                        end
-                      %>
-                      <li class="pub-c-task-list__panel-link <% if panel_link[:active] %>pub-c-task-list__panel-link--active<% end %>">
-                        <a href="<%= panel_link_href %>" class="pub-c-task-list__panel-link-item js-panel-link" data-position="<%= "#{group_index + 1}.#{step_index + 1}.#{panel_index + 1}" %>">
-                          <%= panel_link_text %> <%= panel_link[:text] %>
-                        </a>
-                      </li>
+                          <% if link[:href] %>
+                            </a>
+                          <% end %>
+
+                          <% if link[:cost] %>
+                            <span class="pub-c-task-list__cost"><%= link[:cost] %></span>
+                          <% end %>
+                        </li>
+                      <% end %>
+                    </<%= list_element %>>
+
+                  <% elsif element[:type] == 'substep' %>
+                    <% if in_substep %>
+                      </div>
+                      <% in_substep = false %>
                     <% end %>
-                  </ol>
+                    <div class="pub-c-task-list__substep <% if element[:optional] %>pub-c-task-list__substep--optional<% end %>">
+                    <% in_substep = true %>
+                  <% end %>
+                <% end %>
+
+                <% if in_substep %>
+                  </div>
```

Comment:
Yeah. At this point I'd welcome any suggestion on how to solve this problem in a different way.

Reply:
I'm not really sure what structure you're trying to create?

Reply:
Oh I see! You declare substeps _inline_ with their own content? That seems really weird to me. I'd personally expect the syntax for the config to look like:

```json
{
  type: "substep",
  content: [
     {
       type: "paragraph",
       text: "This paragraph is inside a sub step."
     },
  ]
},
```

Wouldn't that be far more representative of how the data is structured? It would then be trivial to translate this into correctly nested HTML.

Reply:
Substeps, yes. What you suggest would be more representative of the structure, except that substeps are only there as a visual thing, not a data structure thing. In terms of the data they should almost be invisible. Also without partials wouldn't the template code get even more complex and nested using this approach?

Reply:
But they _are_ a data thing, surely? By the time you've started conceptually grouping things into substeps, that is very much part of the data. And if they're visually grouped, I think it makes all the more sense to group them in the config, too. I'd be very much in favour of this approach.

In terms of nesting, yes it would be a little bit unfortunate without partials, but I think it's still vastly preferable over hiding wrapping tags in `if` blocks, which is just confusing and fragile.

Although it's not ideal, you could do as I suggested somewhere else and store your elements in variables/methods at the top of the file, and then write out the rest of the file as if you're using partials, eg:

```erb
<%
  def paragraph(text)
    content_tag(
      :p,
      text,
      class: ... ,
    )
  end

  def list(style, contents)
    tag_name = style == "required" ? "ol" : "ul"
    content_tag(tag_name, ...)
  end

  def substep(...)
    ...
  end
%>

<% step[:contents].each do |element| %>
  <%=  case element[:type]
            when "paragraph"
              paragraph(element[:text])
            when "list"
              list(element[:style], element[:contents])
            when "substep"
              ...
            end %>
<% end %>
```

Reply:
Interesting. Okay, I'll take a look at that.

Reply:
@alecgibson this suggestion is brilliant and I really want to implement it but this all needs to be ready this week and if I rewrite the template like this it'll all need to go back into review and I'll probably introduce some new bugs. I'll definitely look at this approach once it gets into the gem - are you happy for me to leave it as-is for now?

Reply:
Yeah I mean everything's up to you and your team. If you're going to put it in a gem anyway, then maybe it's better to just wait until then; this just would have been a nicer stopgap until then.